pyccel/ast/type_annotations.py

```diff
@@ -269,141 +296,49 @@ class SyntacticTypeAnnotation(PyccelAstNode):
 
     Parameters
     ----------
-    dtypes : list of str
-        The dtypes named in the type annotation.
-
-    ranks : list of int
-        The number of ranks requested for each possible annotation.
+    dtype : PyccelSymbol | IndexedElement | DottedName
+        The dtype named in the type annotation.
 
-    orders : list of str or None
-        The orders requested in the type annotation.
-
-    is_const : bool, optional
-        The constness as specified in the type annotation.
-        If the constness is unknown then its value will be fixed in the
-        semantic stage.
+    order : str | None
+        The order requested in the type annotation.
     """
-    __slots__ = ('_dtypes', '_ranks', '_orders', '_is_const')
+    __slots__ = ('_dtype', '_order')
     _attribute_nodes = ()
-    def __init__(self, dtypes, ranks, orders, is_const = None):
-        if any(not isinstance(d, (str, DottedName)) for d in dtypes):
+    def __init__(self, dtype, order = None):
+        if not isinstance(dtype, (str, DottedName, IndexedElement)):
             raise ValueError("Syntactic datatypes should be strings")
-        if any(not isinstance(r, int) for r in ranks):
-            raise ValueError("Ranks should have integer values")
-        if not all(o is None or isinstance(o, str) for o in orders):
-            raise ValueError("Orders should be strings")
-        if not (isinstance(is_const, bool) or is_const is None):
-            raise ValueError("Is const should be a boolean")
-        self._dtypes = tuple(dtypes)
-        self._ranks = tuple(ranks)
-        self._orders = tuple(o if o != '' else None for o in orders)
-        self._is_const = is_const
```

Comment:
There is no need to check the `_ranks` attribute as it no longer exists in the class. Instead the rank is included in the dtype information. For example `'float[:,:]'` was previously stored as `SyntacticTypeAnnotation(['float'], [2], [None], None)`, but when adding support for `tuple[int,..]` I noticed that this meant I needed to store it as: `SyntacticTypeAnnotation([IndexedElement(PythonTuple, int)], [1], [None], None)` which duplicates the information.

Now `'float[:,:]'` is stored as `SyntacticTypeAnnotation(IndexedElement(PythonFloat, Slice(None, None), Slice(None, None), None, None)`. This will allow us to easily support syntax such as:
```
from pyccel.types import float64
a : float64[:,:]
```
(as suggested in #1487) as the string syntax and the non-string syntax will be equivalent.